File: ios/PttWire/Sources/PttWire/EncryptedHistory.swift
import CryptoKit
import Foundation

/// End-to-end wrapper for persisted SFrame datagrams. The wrapper authenticates packet ordering
/// and routing headers while the object store sees only ciphertext.
public enum EncryptedHistory {
    private static let magic = Data("PTTH".utf8)
    private static let version: UInt8 = 1
    private static let nonceBytes = 12
    private static let maximumFrames = 1_501
    private static let info = Data("PTT Talk encrypted history v1".utf8)

    public static func seal(
        channelId: UUID,
        talkId: UUID,
        membershipEpoch: Int32,
        kid: UInt64,
        baseKey: Data,
        packets: [Data],
        nonce: Data? = nil
    ) throws -> Data {
        let actualNonce = nonce ?? randomNonce()
        try validate(epoch: membershipEpoch, key: baseKey, packets: packets, nonce: actualNonce)
        var plaintext = Data()
        appendBigEndian(UInt32(packets.count), to: &plaintext)
        packets.forEach { plaintext.append($0) }
        let sealed = try AES.GCM.seal(
            plaintext,
            using: deriveKey(baseKey),
            nonce: AES.GCM.Nonce(data: actualNonce),
            authenticating: aad(channelId, talkId, membershipEpoch, kid)
        )
        var output = magic
        output.append(version)
        output.append(actualNonce)
        output.append(sealed.ciphertext)
        output.append(sealed.tag)
        return output
    }

    public static func open(
        _ blob: Data,
        channelId: UUID,
        talkId: UUID,
        membershipEpoch: Int32,
        kid: UInt64,
        baseKey: Data
    ) throws -> [Data] {
        guard blob.count >= magic.count + 1 + nonceBytes + 4 + 16,
              blob.prefix(magic.count) == magic else { throw EncryptedHistoryError.invalidObject }
        guard blob[magic.count] == version else { throw EncryptedHistoryError.unsupportedVersion }
        guard membershipEpoch > 0, baseKey.count == 32 else { throw EncryptedHistoryError.invalidMetadata }
        let nonceStart = magic.count + 1
        let nonce = blob.subdata(in: nonceStart..<(nonceStart + nonceBytes))
        let sealed = blob.suffix(from: nonceStart + nonceBytes)
        guard sealed.count >= 16 else { throw EncryptedHistoryError.invalidObject }
        let box = try AES.GCM.SealedBox(
            nonce: AES.GCM.Nonce(data: nonce),
            ciphertext: sealed.dropLast(16),
            tag: sealed.suffix(16)
        )
        let plaintext = try AES.GCM.open(
            box,
            using: deriveKey(baseKey),
            authenticating: aad(channelId, talkId, membershipEpoch, kid)
        )
        guard plaintext.count >= 4 else { throw EncryptedHistoryError.invalidObject }
        let count = Int(readUInt32(plaintext, at: 0))
        guard (1...maximumFrames).contains(count),
              plaintext.count == 4 + count * productionMediaDatagramBytes
        else { throw EncryptedHistoryError.invalidObject }
        return (0..<count).map { index in
            let start = 4 + index * productionMediaDatagramBytes
            return plaintext.subdata(in: start..<(start + productionMediaDatagramBytes))
        }
    }

    private static func validate(epoch: Int32, key: Data, packets: [Data], nonce: Data) throws {
        guard epoch > 0, key.count == 32 else { throw EncryptedHistoryError.invalidMetadata }
        guard nonce.count == nonceBytes else { throw EncryptedHistoryError.invalidNonce }
        guard (1...maximumFrames).contains(packets.count),
              packets.allSatisfy({ $0.count == productionMediaDatagramBytes })
        else { throw EncryptedHistoryError.invalidObject }
    }

    private static func aad(_ channelId: UUID, _ talkId: UUID, _ epoch: Int32, _ kid: UInt64) -> Data {
        var output = magic
        output.append(version)
        output.append(contentsOf: PttWire.uuidBytes(channelId))
        output.append(contentsOf: PttWire.uuidBytes(talkId))
        appendBigEndian(UInt32(bitPattern: epoch), to: &output)
        appendBigEndian(kid, to: &output)
        return output
    }

    private static func deriveKey(_ input: Data) -> SymmetricKey {
        HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: input),
            salt: Data(repeating: 0, count: 32),
            info: info,
            outputByteCount: 32
        )
    }

    private static func randomNonce() -> Data {
        var bytes = [UInt8](repeating: 0, count: nonceBytes)
        precondition(SecRandomCopyBytes(kSecRandomDefault, nonceBytes, &bytes) == errSecSuccess)
        return Data(bytes)
    }
}

public enum EncryptedHistoryError: Error, Equatable {
    case invalidObject
    case unsupportedVersion
    case invalidMetadata
    case invalidNonce
}

private func appendBigEndian<T: FixedWidthInteger>(_ value: T, to data: inout Data) {
    var encoded = value.bigEndian
    data.append(Data(bytes: &encoded, count: MemoryLayout<T>.size))
}

private func readUInt32(_ data: Data, at offset: Int) -> UInt32 {
    data[offset..<(offset + 4)].reduce(0) { ($0 << 8) | UInt32($1) }
}

File: android/media/src/main/kotlin/app/ptt/media/EncryptedHistory.kt
package app.ptt.media

import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import java.util.UUID
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * End-to-end encrypted, self-delimiting history object. The server stores this blob verbatim and
 * never receives the media epoch key. Wrapping the already-SFrame-protected datagrams also binds
 * their ordering and routing headers, which are otherwise authenticated only at the relay hop.
 */
object EncryptedHistory {
    private val MAGIC = byteArrayOf('P'.code.toByte(), 'T'.code.toByte(), 'T'.code.toByte(), 'H'.code.toByte())
    private const val VERSION = 1
    private const val NONCE_BYTES = 12
    private const val TAG_BITS = 128
    private const val MAX_FRAMES = 1_501
    private val INFO = "PTT Talk encrypted history v1".toByteArray(StandardCharsets.UTF_8)

    fun seal(
        channelId: UUID,
        talkId: UUID,
        membershipEpoch: Int,
        kid: ULong,
        baseKey: ByteArray,
        packets: List<ByteArray>,
        nonce: ByteArray = ByteArray(NONCE_BYTES).also(SecureRandom()::nextBytes),
    ): ByteArray {
        validate(membershipEpoch, baseKey, packets, nonce)
        val prefix = MAGIC + byteArrayOf(VERSION.toByte()) + nonce
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(deriveKey(baseKey), "AES"), GCMParameterSpec(TAG_BITS, nonce))
        cipher.updateAAD(aad(channelId, talkId, membershipEpoch, kid))
        val plaintext = ByteBuffer.allocate(4 + packets.size * MEDIA_DATAGRAM_BYTES)
            .putInt(packets.size)
            .also { output -> packets.forEach(output::put) }
            .array()
        return prefix + cipher.doFinal(plaintext)
    }

    fun open(
        blob: ByteArray,
        channelId: UUID,
        talkId: UUID,
        membershipEpoch: Int,
        kid: ULong,
        baseKey: ByteArray,
    ): List<ByteArray> {
        require(blob.size >= MAGIC.size + 1 + NONCE_BYTES + 4 + TAG_BITS / 8) { "history object is truncated" }
        require(blob.copyOfRange(0, MAGIC.size).contentEquals(MAGIC)) { "invalid history magic" }
        require(blob[MAGIC.size].toInt() and 0xff == VERSION) { "unsupported history version" }
        require(membershipEpoch > 0 && baseKey.size == 32) { "invalid history key metadata" }
        val nonceStart = MAGIC.size + 1
        val nonce = blob.copyOfRange(nonceStart, nonceStart + NONCE_BYTES)
        val ciphertext = blob.copyOfRange(nonceStart + NONCE_BYTES, blob.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(deriveKey(baseKey), "AES"), GCMParameterSpec(TAG_BITS, nonce))
        cipher.updateAAD(aad(channelId, talkId, membershipEpoch, kid))
        val plaintext = cipher.doFinal(ciphertext)
        require(plaintext.size >= 4) { "history plaintext is truncated" }
        val input = ByteBuffer.wrap(plaintext)
        val count = input.int
        require(count in 1..MAX_FRAMES) { "invalid history frame count" }
        require(input.remaining() == count * MEDIA_DATAGRAM_BYTES) { "invalid history object length" }
        return List(count) { ByteArray(MEDIA_DATAGRAM_BYTES).also(input::get) }
    }

    private fun validate(epoch: Int, key: ByteArray, packets: List<ByteArray>, nonce: ByteArray) {
        require(epoch > 0) { "membership epoch must be positive" }
        require(key.size == 32) { "media base key must contain 32 bytes" }
        require(nonce.size == NONCE_BYTES) { "history nonce must contain 12 bytes" }
        require(packets.size in 1..MAX_FRAMES) { "history must contain 1..$MAX_FRAMES frames" }
        require(packets.all { it.size == MEDIA_DATAGRAM_BYTES }) { "history packets must contain 160 bytes" }
    }

    private fun aad(channelId: UUID, talkId: UUID, epoch: Int, kid: ULong): ByteArray =
        ByteBuffer.allocate(4 + 1 + 16 + 16 + 4 + 8)
            .put(MAGIC)
            .put(VERSION.toByte())
            .putLong(channelId.mostSignificantBits)
            .putLong(channelId.leastSignificantBits)
            .putLong(talkId.mostSignificantBits)
            .putLong(talkId.leastSignificantBits)
            .putInt(epoch)
            .putLong(kid.toLong())
            .array()

    private fun deriveKey(inputKey: ByteArray): ByteArray {
        val extract = Mac.getInstance("HmacSHA256").apply {
            init(SecretKeySpec(ByteArray(32), "HmacSHA256"))
        }.doFinal(inputKey)
        return Mac.getInstance("HmacSHA256").apply {
            init(SecretKeySpec(extract, "HmacSHA256"))
        }.doFinal(INFO + byteArrayOf(1))
    }
}
